README.md
# Docker container images with "headless" VNC session

This repository contains a collection of Docker images with headless VNC environments.

Each Docker image is installed with the following components:

* Desktop environment [**IceWM**](http://www.icewm.org/)
* VNC-Server (default VNC port `5901`)
* [**noVNC**](https://github.com/novnc/noVNC) - HTML5 VNC client (default http port `6901`)
* Browsers:
  * Mozilla Firefox
  * Chromium
  
![Docker VNC Desktop access via HTML page](.pics/vnc_container_view.png)

## Build Status
`master`:  [![Build Status MASTER](https://travis-ci.org/ConSol/docker-headless-vnc-container.svg?branch=master)](https://travis-ci.org/ConSol/docker-headless-vnc-container) `dev`: [![Build Status DEV](https://travis-ci.org/ConSol/docker-headless-vnc-container.svg?branch=dev)](https://travis-ci.org/ConSol/docker-headless-vnc-container)

## Current provided OS & UI sessions:
* `baztian/headless-vnc-container`: __Ubuntu with `IceWM` UI session__

  [![](https://images.microbadger.com/badges/version/baztian/headless-vnc-container.svg)](https://hub.docker.com/r/baztian/headless-vnc-container/) [![](https://images.microbadger.com/badges/image/baztian/headless-vnc-container.svg)](http://microbadger.com/images/baztian/headless-vnc-container)

## OpenShift / Kubernetes

It's also possible to run the images in container orchestration platforms like [Kubernetes](https://kubernetes.io) or [OpenShift](https://openshift.io/). For more information how to deploy containers in the cluster, take a look at:

* [Kubernetes usage of "headless" VNC Docker images](./kubernetes/README.md)
* [OpenShift usage of "headless" VNC Docker images](./openshift/README.md) 

## Usage
Usage is **similar** for all provided images, e.g. for `baztian/headless-vnc-container`:

- Print out help page:

      docker run baztian/headless-vnc-container --help

- Run command with mapping to local port `5901` (vnc protocol) and `6901` (vnc web access):

      docker run -d -p 5901:5901 -p 6901:6901 baztian/headless-vnc-container
  
- Change the default user and group within a container to your own with adding `--user $(id -u):$(id -g)`:

      docker run -d -p 5901:5901 -p 6901:6901 --user $(id -u):$(id -g) baztian/headless-vnc-container

- If you want to get into the container use interactive mode `-it` and `bash`
      
      docker run -it -p 5901:5901 -p 6901:6901 baztian/headless-vnc-container bash

- Build an image from scratch:

      docker build -t baztian/headless-vnc-container headless-vnc-container

# Connect & Control
If the container is started like mentioned above, connect via one of these options:

* connect via __VNC viewer `localhost:5901`__, default password: `vncpassword`
* connect via __noVNC HTML5 full client__: [`http://localhost:6901/vnc.html`](http://localhost:6901/vnc.html), default password: `vncpassword` 
* connect via __noVNC HTML5 lite client__: [`http://localhost:6901/?password=vncpassword`](http://localhost:6901/?password=vncpassword) 


## Hints

### 1) Extend a Image with your own software
Since version `1.1.0` all images run as non-root user per default, so if you want to extend the image and install software, you have to switch back to the `root` user:

```bash
## Custom Dockerfile
FROM baztian/headless-vnc-container
ENV REFRESHED_AT 2018-03-18

# Switch to root user to install additional software
USER 0

## Install a gedit
RUN yum install -y gedit \
    && yum clean all

## switch back to default user
USER 1000
```

### 2) Change User of running Sakuli Container

Per default, since version `1.3.0` all container processes will be executed with user id `1000`. You can change the user id as follows: 

#### 2.1) Using root (user id `0`)
Add the `--user` flag to your docker run command:

    docker run -it --user 0 -p 6911:6901 baztian/headless-vnc-container

#### 2.2) Using user and group id of host system
Add the `--user` flag to your docker run command:

    docker run -it -p 6911:6901 --user $(id -u):$(id -g) baztian/headless-vnc-container

### 3) Override VNC environment variables
The following VNC environment variables can be overwritten at the `docker run` phase to customize your desktop environment inside the container:
* `VNC_COL_DEPTH`, default: `24`
* `VNC_RESOLUTION`, default: `1280x1024`
* `VNC_PW`, default: `my-pw`

#### 3.1) Example: Override the VNC password
Simply overwrite the value of the environment variable `VNC_PW`. For example in
the docker run command:

    docker run -it -p 5901:5901 -p 6901:6901 -e VNC_PW=my-pw baztian/headless-vnc-container

#### 3.2) Example: Override the VNC resolution
Simply overwrite the value of the environment variable `VNC_RESOLUTION`. For example in
the docker run command:

    docker run -it -p 5901:5901 -p 6901:6901 -e VNC_RESOLUTION=800x600 baztian/headless-vnc-container
    
### 4) View only VNC
Since version `1.2.0` it's possible to prevent unwanted control via VNC. Therefore you can set the environment variable `VNC_VIEW_ONLY=true`. If set, the startup script will create a random password for the control connection and use the value of `VNC_PW` for view only connection over the VNC connection.

     docker run -it -p 5901:5901 -p 6901:6901 -e VNC_VIEW_ONLY=true baztian/headless-vnc-container

### 5) Known Issues

#### 5.1) Chromium crashes with high VNC_RESOLUTION ([#53](https://github.com/ConSol/docker-headless-vnc-container/issues/53))
If you open some graphic/work intensive websites in the Docker container (especially with high resolutions e.g. `1920x1080`) it can happen that Chromium crashes without any specific reason. The problem there is the too small `/dev/shm` size in the container. Currently there is no other way, as define this size on startup via `--shm-size` option, see [#53 - Solution](https://github.com/ConSol/docker-headless-vnc-container/issues/53#issuecomment-347265977):

    docker run --shm-size=256m -it -p 6901:6901 -e VNC_RESOLUTION=1920x1080 baztian/headless-vnc-container chromium-browser http://map.norsecorp.com/
  
Thx @raghavkarol for the hint! 

## How to release
See **[how-to-release.md](./how-to-release.md)**

## Contributors

At this point we want to thank all contributors, which helped to move this great project by submitting code, writing documentation, or adapting other tools to play well together with the docker headless container.

* [Tobias Schneck](https://github.com/toschneck) - Lead development
* [Robert Bohne](https://github.com/rbo) - IceWM images
* [hsiaoyi0504](https://github.com/hsiaoyi0504) - PR [#66](https://github.com/ConSol/docker-headless-vnc-container/pull/66)
* [dmhumph](https://github.com/dmhumph) - PR [#44](https://github.com/ConSol/docker-headless-vnc-container/issue/44) 
* [Simon Hofmann](https://github.com/s1hofmann) - Current maintainer

## Changelog

The current changelog is provided here: **[changelog.md](./changelog.md)**

## Contact
For questions, professional support or maybe some hints, feel free to contact us via **[testautomatisierung@consol.de](mailto:testautomatisierung@consol.de)** or open an [issue](https://github.com/ConSol/docker-headless-vnc-container/issues/new).

The guys behind:

**ConSol Software GmbH** <br/>
*Franziskanerstr. 38, D-81669 Munich* <br/>
*Tel. +49-89-45841-100, Fax +49-89-45841-111*<br/>
*Homepage: http://www.consol.de E-Mail: [info@consol.de](info@consol.de)*


changelog.md
## Changelog of the Docker headless VNC images

* Next version - unreleased
  * Fix documentation

* 2.4.3 - 2020-03-12
  * Fix wrong chromium sandobx parameter in all relevant places

* 2.4.2 - 2020-01-27
  * Fix wrong chromium sandobx parameter

* 2.4.1 - 2020-01-27
  * Fix firefox promoting updates

* 2.4.0 - 2020-01-24
  * Upgrade novnc

* 2.3.0 - 2020-01-24
  * TigerVNC

* 2.2.0 - 2020-01-24
  * Upgrade Firefox

* 2.1.0 - 2020-01-24
  * Upgrade to Ubuntu 18.04
* 2.0.1 - 2020-01-24
  * fixed icewm startup

* 2.0.0 - 2020-01-15
  * removed CentOS support
  * removed Xfce support
  
* 1.4.0 - 2018-10-29
  * added Github issue and pull request templates
  * added install script for custom fonts ([#83](https://github.com/ConSol/docker-headless-vnc-container/issues/83)))
  * fixed problem that VNC password was not properly set after doing a Docker commit ([#88](https://github.com/ConSol/  docker-headless-vnc-container/issues/88))
  * fixed some typos ([#87](https://github.com/ConSol/docker-headless-vnc-container/issues/87))
  * added 'how-to-release' info ([#75](https://github.com/ConSol/docker-headless-vnc-container/issues/75))
  * added Travis build ([#74](https://github.com/ConSol/docker-headless-vnc-container/issues/74))
  
* 1.3.0 - 2018-03-27
  * change default USER to `1000` ([#61](https://github.com/ConSol/docker-headless-vnc-container/issues/61))
  * refactor vnc startup script ([#73](https://github.com/ConSol/docker-headless-vnc-container/issues/73))
    * add help option `--help`
    * ensure correct forwarding of shutdown signals
    * add "DEBUG" mode and move all log output to this mode
    * update README.md
  * merge pull request from:
    * [hsiaoyi0504](https://github.com/hsiaoyi0504) update noVNC to [v1.0.0](https://github.com/novnc/noVNC/releases/  tag/v1.0.0) ([#66](https://github.com/ConSol/docker-headless-vnc-container/pull/66))
  * add example for [Kubernetes usage](./kubernetes/README.md) ([#71](https://github.com/ConSol/  docker-headless-vnc-container/issues/71)) 
  * remove verbose output by default from `set_user_permissions.sh` script
   
* 1.2.3 - 2017-12-15
  * start no_vnc with localhost instead of VNC_IP to ensure startup on docker bridge mode ([#41](https://github.com/  ConSol/docker-headless-vnc-container/issue/41)) 
  * add xfce4-terminal to enable copy and paste ([#55](https://github.com/ConSol/docker-headless-vnc-container/issue/  55))
  * add save mode for tag_image.sh script 
  * disable trigger latest Dockerhub build -> use tag_image.sh script 
  * use correct python-numpy packages 
  * add "Known Issues" documentation: [Chromium crashes with high VNC_RESOLUTION](https://github.com/ConSol/  docker-headless-vnc-container#51-chromium-crashes-with-high-vnc_resolution-53)
  
* 1.2.2 - 2017-11-20
  * disable automatic firefox updates ([#45](https://github.com/ConSol/docker-headless-vnc-container/issue/45))
  * fix wrong env usage of `$HOME` ([#47](https://github.com/ConSol/docker-headless-vnc-container/issue/47))
  * fix hanging vnc handshake if container is offline ([#50](https://github.com/ConSol/docker-headless-vnc-container/  issue/50))
  * Merge pull request from:
      * [dmhumph](https://github.com/dmhumph): Update OpenShift instructions for OCP 3.5 or greater ([#44](https://  github.com/ConSol/docker-headless-vnc-container/issue/44)) 
  
* 1.2.1 - 2017-09-27
  * Centos7: use temporally epel-testing repo to fix chrome error `libpng warning: Application built with   libpng-1.6.22 but running with 1.5.13` in version `60.0.3112.113-2.el7`
  * Fix broken `docker stop/start` ([#40](https://github.com/ConSol/docker-headless-vnc-container/issues/40))
  
* 1.2.0 - 2017-09-18
  * Use fixed firefox version `45.9.0esr` ([#39](https://github.com/ConSol/docker-headless-vnc-container/issues/39),   [#20](https://github.com/ConSol/docker-headless-vnc-container/issues/20), [#16](https://github.com/ConSol/  docker-headless-vnc-container/issues/16)) 
    * fix memory consumption problems in comparison to current firefox 55
    * tested on long term container issues without crashes 
  * Add en_US.UTF-8 as default locale ([#27](https://github.com/ConSol/docker-headless-vnc-container/issues/27)) 
  * Add enviroment variable `VNC_VIEW_ONLY=true` for view only VNC connections ([#25](https://github.com/ConSol/  docker-headless-vnc-container/issues/25))
    * Update to tigervnc version `1.8`
  * Use a more stable command for ip determination (PR [#31](https://github.com/ConSol/docker-headless-vnc-container/  issues/31), thx to @arnaudveron) 
  * Support empty `CMD` values in startup script ([#32](https://github.com/ConSol/docker-headless-vnc-container/  issues/32)) 
  * Add chromium flag `--disable-gpu` to work for current chromium ([#36](https://github.com/ConSol/  docker-headless-vnc-container/issues/36)) 
  * Add `TERM=xterm` environment var ([#37](https://github.com/ConSol/docker-headless-vnc-container/issues/37))  
  
* 1.1.0 - 2017-03-18
  * Add OpenShift support ([#3](https://github.com/ConSol/docker-headless-vnc-container/issues/3))
     * Support for start container process as non-root in usermode ([#7](https://github.com/ConSol/  docker-headless-vnc-container/issues/7))
     * Add Labels for OpenShift
  * Add [IceWM](http://www.icewm.org/) as alternative UI
  * Remove Java installation, not needed in this image ([#5](https://github.com/ConSol/docker-headless-vnc-container/  issues/5))
  * Refactor build structure ([#5](https://github.com/ConSol/docker-headless-vnc-container/issues/5))
  * Fix chromium/chrome startup ([#2](https://github.com/ConSol/docker-headless-vnc-container/issues/2)): Set   chromium as default "chrome" and support `~/.chromium-browser.init`
  * Switch from vnc4server to tigervnc ([#4](https://github.com/ConSol/docker-headless-vnc-container/issues/4))
  * Remove not needed Screensaver pacakges ([#11](https://github.com/ConSol/docker-headless-vnc-container/issues/11))          

src/install/firefox.sh
#!/usr/bin/env bash
set -e

echo "Install Firefox"

function disableUpdate(){
    ff_def="$1/browser/defaults/profile"
    mkdir -p $ff_def
    cat > $ff_def/user.js <<EOF_FF 
user_pref("browser.startup.homepage", "about:blank");
user_pref("datareporting.policy.firstRunURL", "");
user_pref("app.normandy.firstRun", false);
user_pref("app.update.auto", false);
user_pref("app.update.enabled", false);
user_pref("app.update.lastUpdateTime.addon-background-update-timer", 1182011519);
user_pref("app.update.lastUpdateTime.background-update-timer", 1182011519);
user_pref("app.update.lastUpdateTime.blocklist-background-update-timer", 1182010203);
user_pref("app.update.lastUpdateTime.microsummary-generator-update-timer", 1222586145);
user_pref("app.update.lastUpdateTime.search-engine-update-timer", 1182010203);
EOF_FF
}

#copy from org/sakuli/common/bin/installer_scripts/linux/install_firefox_portable.sh
function instFF() {
    if [ ! "${1:0:1}" == "" ]; then
        FF_VERS=$1
        if [ ! "${2:0:1}" == "" ]; then
            FF_INST=$2
            echo "download Firefox $FF_VERS and install it to '$FF_INST'."
            mkdir -p "$FF_INST"
            FF_URL=http://releases.mozilla.org/pub/firefox/releases/$FF_VERS/linux-x86_64/en-US/firefox-$FF_VERS.tar.bz2
            echo "FF_URL: $FF_URL"
            wget -qO- $FF_URL | tar xvj --strip 1 -C $FF_INST/
            ln -s "$FF_INST/firefox" /usr/bin/firefox
            disableUpdate $FF_INST
            exit $?
        fi
    fi
    echo "function parameter are not set correctly please call it like 'instFF [version] [install path]'"
    exit -1
}

instFF '68.4.2esr' '/usr/lib/firefox'

#apt-get update
#apt-get install -y firefox
#apt-get install -y firefox=45*
#apt-mark hold firefox
#apt-get clean -y

.build/trigger_dockerhub.sh
#!/usr/bin/env bash
set -e -o pipefail


if [ -z $GIT_BRANCH ] ; then
    GIT_BRANCH=$TRAVIS_BRANCH
fi
if [ -z $GIT_BRANCH ] ; then
    echo "cloud not execute script! Please specify at least on environment variable of: GIT_BRANCH, TRAVIS_BRANCH"
    exit -1
fi

echo "branch=$GIT_BRANCH"
GIT_BRANCH=${GIT_BRANCH/origin\/}
DOCKER_TAG="${GIT_BRANCH/refs\/tags\/}"

if [[ $DOCKER_TAG == "master" ]] ; then
   echo "skip building latest tag!"
   echo "... use 'tag_image.sh' script to release a new version. See: https://github.com/ConSol/docker-headless-vnc-container/blob/master/how-to-release.md"
   exit 0
fi

echo "DOCKER_TAG=$DOCKER_TAG"
echo "..."
echo "trigger dockerhub builds for Tag $DOCKER_TAG:"

URLS=(
	"https://registry.hub.docker.com/u/consol/ubuntu-xfce-vnc/trigger/bfc5c145-bddf-404c-8ae7-356f928656e9/"
    "https://registry.hub.docker.com/u/consol/ubuntu-icewm-vnc/trigger/c73f3c3f-ecc3-46b1-9766-46c7dfca4d16/"
)
PAYLOAD='{"source_type": "Tag", "source_name": "'$DOCKER_TAG'"}'

# use docker tag instead of branch
if [[ $DOCKER_TAG == "dev" ]] ; then
   PAYLOAD='{"docker_tag": "'dev'"}'
fi

# use first parameter to filter trigger command
IMAGENAME=$1

#Loop
for URL in "${URLS[@]}"
do
    if [ -z $IMAGENAME ] || [[ $URL =~ .*"$IMAGENAME".* ]] ; then
        echo "URL: $URL"
        echo "PAYLOAD: $PAYLOAD"
        curl -H "Content-Type: application/json" --data "$PAYLOAD" -X POST "$URL"
        echo " - done!"
    fi
done

src/scripts/vnc_startup.sh
#!/bin/bash
### every exit != 0 fails the script
set -e

## print out help
help (){
echo "
USAGE:
docker run -it -p 6901:6901 -p 5901:5901 baztian/<image>:<tag> <option>

IMAGES:
baztian/headless-vnc-container

TAGS:
latest  stable version of branch 'master'
dev     current development version of branch 'dev'

OPTIONS:
-w, --wait      (default) keeps the UI and the vncserver up until SIGINT or SIGTERM will received
-s, --skip      skip the vnc startup and just execute the assigned command.
                example: docker run baztian/headless-vnc-container --skip bash
-d, --debug     enables more detailed startup output
                e.g. 'docker run baztian/headless-vnc-container --debug bash'
-h, --help      print out this help

Fore more information see: https://github.com/baztian/docker-headless-vnc-container
"
}
if [[ $1 =~ -h|--help ]]; then
    help
    exit 0
fi

# should also source $STARTUPDIR/generate_container_user
source $HOME/.bashrc

# add `--skip` to startup args, to skip the VNC startup procedure
if [[ $1 =~ -s|--skip ]]; then
    echo -e "\n\n------------------ SKIP VNC STARTUP -----------------"
    echo -e "\n\n------------------ EXECUTE COMMAND ------------------"
    echo "Executing command: '${@:2}'"
    exec "${@:2}"
fi
if [[ $1 =~ -d|--debug ]]; then
    echo -e "\n\n------------------ DEBUG VNC STARTUP -----------------"
    export DEBUG=true
fi

## correct forwarding of shutdown signal
cleanup () {
    kill -s SIGTERM $!
    exit 0
}
trap cleanup SIGINT SIGTERM

## write correct window size to chrome properties
$STARTUPDIR/chrome-init.sh

## resolve_vnc_connection
VNC_IP=$(hostname -i)

## change vnc password
echo -e "\n------------------ change VNC password  ------------------"
# first entry is control, second is view (if only one is valid for both)
mkdir -p "$HOME/.vnc"
PASSWD_PATH="$HOME/.vnc/passwd"

if [[ -f $PASSWD_PATH ]]; then
    echo -e "\n---------  purging existing VNC password settings  ---------"
    rm -f $PASSWD_PATH
fi

if [[ $VNC_VIEW_ONLY == "true" ]]; then
    echo "start VNC server in VIEW ONLY mode!"
    #create random pw to prevent access
    echo $(head /dev/urandom | tr -dc A-Za-z0-9 | head -c 20) | vncpasswd -f > $PASSWD_PATH
fi
echo "$VNC_PW" | vncpasswd -f >> $PASSWD_PATH
chmod 600 $PASSWD_PATH


## start vncserver and noVNC webclient
echo -e "\n------------------ start noVNC  ----------------------------"
if [[ $DEBUG == true ]]; then echo "$NO_VNC_HOME/utils/launch.sh --vnc localhost:$VNC_PORT --listen $NO_VNC_PORT"; fi
$NO_VNC_HOME/utils/launch.sh --vnc localhost:$VNC_PORT --listen $NO_VNC_PORT &> $STARTUPDIR/no_vnc_startup.log &
PID_SUB=$!

echo -e "\n------------------ start VNC server ------------------------"
echo "remove old vnc locks to be a reattachable container"
vncserver -kill $DISPLAY &> $STARTUPDIR/vnc_startup.log \
    || rm -rfv /tmp/.X*-lock /tmp/.X11-unix &> $STARTUPDIR/vnc_startup.log \
    || echo "no locks present"

echo -e "start vncserver with param: VNC_COL_DEPTH=$VNC_COL_DEPTH, VNC_RESOLUTION=$VNC_RESOLUTION\n..."
if [[ $DEBUG == true ]]; then echo "vncserver $DISPLAY -depth $VNC_COL_DEPTH -geometry $VNC_RESOLUTION"; fi
vncserver $DISPLAY -depth $VNC_COL_DEPTH -geometry $VNC_RESOLUTION &> $STARTUPDIR/no_vnc_startup.log
echo -e "start window manager\n..."
$HOME/wm_startup.sh &> $STARTUPDIR/wm_startup.log

## log connect options
echo -e "\n\n------------------ VNC environment started ------------------"
echo -e "\nVNCSERVER started on DISPLAY= $DISPLAY \n\t=> connect via VNC viewer with $VNC_IP:$VNC_PORT"
echo -e "\nnoVNC HTML client started:\n\t=> connect via http://$VNC_IP:$NO_VNC_PORT/?password=...\n"


if [[ $DEBUG == true ]] || [[ $1 =~ -t|--tail-log ]]; then
    echo -e "\n------------------ $HOME/.vnc/*$DISPLAY.log ------------------"
    # if option `-t` or `--tail-log` block the execution and tail the VNC log
    tail -f $STARTUPDIR/*.log $HOME/.vnc/*$DISPLAY.log
fi

if [ -z "$1" ] || [[ $1 =~ -w|--wait ]]; then
    wait $PID_SUB
else
    # unknown option ==> call command
    echo -e "\n\n------------------ EXECUTE COMMAND ------------------"
    echo "Executing command: '$@'"
    exec "$@"
fi